presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -411,6 +416,76 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
+        ImmutableList.Builder<ColumnMetadata> insertedColumnsBuilder = ImmutableList.builder();
+
+        for (ColumnMetadata column : tableMetadata.getColumns()) {
+            if (column.isHidden()) {
+                continue;
+            }
+            Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
+            int index = refreshMV.getColumns().indexOf(columns.get(column.getName()));
+            if (index < 0) {
+                if (supportsMissingColumnsOnInsert) {
+                    continue;
+                }
+                Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
+                assignments.put(output, cast);
+                insertedColumnsBuilder.add(column);
+            }
+            else {
+                Symbol input = plan.getSymbol(index);
+                Type tableType = column.getType();
+                Type queryType = symbolAllocator.getTypes().get(input);
+
+                if (queryType.equals(tableType) || typeCoercion.isTypeOnlyCoercion(queryType, tableType)) {
+                    assignments.put(output, input.toSymbolReference());
+                }
+                else {
+                    Expression cast = noTruncationCast(input.toSymbolReference(), queryType, tableType);
+                    assignments.put(output, cast);
+                }
+                insertedColumnsBuilder.add(column);
+            }
+        }
+
+        ProjectNode projectNode = new ProjectNode(idAllocator.getNextId(), plan.getRoot(), assignments.build());
+
+        List<ColumnMetadata> insertedColumns = insertedColumnsBuilder.build();
+        List<Field> fields = insertedColumns.stream()
+                .map(column -> Field.newUnqualified(column.getName(), column.getType()))
+                .collect(toImmutableList());
+        Scope scope = Scope.builder().withRelationType(RelationId.anonymous(), new RelationType(fields)).build();
+
+        plan = new RelationPlan(projectNode, scope, projectNode.getOutputSymbols(), Optional.empty());
+
+        String catalogName = refreshMV.getTarget().getCatalogName().getCatalogName();
+        TableStatisticsMetadata statisticsMetadata = metadata.getStatisticsCollectionMetadataForWrite(session, catalogName, tableMetadata.getMetadata());
+
+        List<String> insertedTableColumnNames = insertedColumns.stream()
+                .map(ColumnMetadata::getName)
+                .collect(toImmutableList());
+
+        Optional<NewTableLayout> newTableLayout = metadata.getInsertLayout(session, refreshMV.getTarget());
+        return createTableWriterPlan(
+            analysis,
+            plan,
+            new TableWriterNode.RefreshMaterializedViewReference(refreshMV.getMaterializedViewHandle(), refreshMV.getTarget(), new ArrayList<>(analysis.getTables())),
+            insertedTableColumnNames,
+            newTableLayout,
+            statisticsMetadata);
```

Comment:
This looks exactly the same as `createInsertPlan`, except for that target table handles and a couple of other things. There's probably some opportunity for reuse, especially given the complexity of this method.

Reply:
I did some refactoring .. let me know what you think.